include/util/multi_level_partition.hpp

```diff
@@ -1,31 +1,310 @@
+        partition.resize(num_nodes + 1, 0);
+        NodeID sentinel = num_nodes;
+
+        // Sort nodes bottum-up by cell id.
+        // This ensures that we get a nice grouping from parent to child cells:
+        //
+        // intitial:
+        // level 0: 0 1 2 3 4 5
+        // level 1: 2 1 3 4 3 4
+        // level 2: 2 2 0 1 0 1
+        //
+        // first round:
+        // level 0: 1 0 2 4 3 5
+        // level 1: 1 2 3 3 4 4 (< sorted)
+        // level 2: 2 2 0 0 1 1
+        //
+        // second round:
+        // level 0: 2 4 3 5 1 0
+        // level 1: 3 3 4 4 1 2
+        // level 2: 0 0 1 1 2 2 (< sorted)
+        for (const auto &partition : partitions)
+        {
+            std::stable_sort(permutation.begin(),
+                             permutation.end(),
+                             [&partition](const auto lhs, const auto rhs) {
+                                 return partition[lhs] < partition[rhs];
+                             });
+        }
+
+        // top down assign new cell ids
+        LevelID level = partitions.size();
+        for (const auto &partition : boost::adaptors::reverse(partitions))
+        {
+            BOOST_ASSERT(permutation.size() > 0);
+            CellID last_cell_id = partition[permutation.front()];
+            CellID cell_id = 0;
+            for (const auto node : permutation)
+            {
+                if (last_cell_id != partition[node])
+                {
+                    cell_id++;
+                    last_cell_id = partition[node];
+                }
+                SetCellID(level, node, cell_id);
+            }
+            // Store the number of cells of the level in the sentinel
+            SetCellID(level, sentinel, cell_id + 1);
+            level--;
+        }
+
+        // level 1 does not have child cells
+        level_to_children_offset.push_back(0);
+
+        for (auto level_idx = 0UL; level_idx < partitions.size() - 1; ++level_idx)
+        {
+            const auto &parent_partition = partitions[level_idx + 1];
+
+            level_to_children_offset.push_back(cell_to_children.size());
+
+            CellID last_parent_id = parent_partition[permutation.front()];
+            cell_to_children.push_back(GetCell(level_idx + 1, permutation.front()));
+            for (const auto node : permutation)
+            {
+                if (last_parent_id != parent_partition[node])
+                {
+                    // Note: we use the new cell id here, not the ones contained
+                    // in the input partition
+                    cell_to_children.push_back(GetCell(level_idx + 1, node));
+                    last_parent_id = parent_partition[node];
+                }
+            }
+            // insert sentinel for the last cell
+            cell_to_children.push_back(GetCell(level_idx + 1, permutation.back()) + 1);
+        }
+    }
+
+    std::vector<PartitionID> partition;
+    std::vector<std::size_t> level_offsets;
```

Comment:
Here better to use fixed-size `uint8_t` with assumption of no more 255 bits per level, atm it is 63